packages/controls/src/widget_tagsinput.ts

```diff
@@ -0,0 +1,844 @@
+  }
+
+  abstract createTag(value: any, index: number, selected: boolean): HTMLElement;
+  abstract updateTag(
+    tag: HTMLElement,
+    value: any,
+    index: number,
+    selected: boolean
+  ): void;
+
+  el: HTMLDivElement;
+  taginputWrapper: HTMLDivElement;
+  taginput: HTMLInputElement;
+  autocompleteList: HTMLDataListElement;
+  tags: HTMLElement[];
+  hoveredTag: HTMLElement | null = null;
+  hoveredTagIndex: number | null = null;
+  datalistID: string;
+  inputIndex: number;
+  selection: null | Selection;
+  preventLoosingFocus: boolean;
+
+  model: TagsInputBaseModel;
+}
+
+export class TagsInputModel extends TagsInputBaseModel {
+  defaults(): Backbone.ObjectHash {
+    return {
+      ...super.defaults(),
+      value: [],
+      tag_style: '',
+      _view_name: 'TagsInputView',
+      _model_name: 'TagsInputModel'
+    };
+  }
+}
+
+export class TagsInputView extends TagsInputBaseView {
+  /**
+   * Create the string tag
+   */
+  createTag(value: string, index: number, selected: boolean): HTMLDivElement {
+    const tag = document.createElement('div');
+    const style: string = this.model.get('tag_style');
+
+    tag.classList.add('jupyter-widget-tag');
+    tag.classList.add(TagsInputView.class_map[style]);
+
+    if (selected) {
+      tag.classList.add('mod-active');
+    }
+
+    tag.appendChild(document.createTextNode(this.getTagText(value)));
+
+    const i = document.createElement('i');
+    i.classList.add('fa');
+    i.classList.add('fa-times');
+    i.classList.add('jupyter-widget-tag-close');
+    tag.appendChild(i);
+
+    i.onmousedown = ((index: number) => {
+      return () => {
+        this.removeTag(index);
+        this.loseFocus();
+      };
+    })(index);
+
+    return tag;
+  }
+
+  /**
+   * Returns the text that should be displayed in the tag element
+   */
+  getTagText(value: string) {
+    return value;
+  }
+
+  /**
+   * Update a given tag
+   */
+  updateTag(
+    tag: HTMLDivElement,
+    value: any,
+    index: number,
+    selected: boolean
+  ): void {
+    if (selected) {
+      tag.classList.add('mod-active');
+    } else {
+      tag.classList.remove('mod-active');
+    }
+  }
+
+  model: TagsInputModel;
+
+  static class_map: Dict<string> = {
+    primary: 'mod-primary',
+    success: 'mod-success',
+    info: 'mod-info',
+    warning: 'mod-warning',
+    danger: 'mod-danger'
+  };
+}
+
+export class ColorsInputModel extends TagsInputBaseModel {
+  defaults(): Backbone.ObjectHash {
+    return {
+      ...super.defaults(),
+      value: [],
+      _view_name: 'ColorsInputView',
+      _model_name: 'ColorsInputModel'
+    };
+  }
+}
+
+export class ColorsInputView extends TagsInputBaseView {
+  /**
+   * Create the Color tag
+   */
+  createTag(value: string, index: number, selected: boolean): HTMLDivElement {
+    const tag = document.createElement('div');
+    const color = value;
+    const darkerColor: string = d3Color
+      .color(value)!
+      .darker()
+      .toString();
```

Comment:
Ping @martinRenou I had to add `!` because `d3Color.color(value)` may be null.

Thoughts?

Reply:
Hmm I don't understand that code, does this actually compile?

Reply:
Yes, it returns `null` if `d3Color.color(value)` is null.

Reply:
What is `d3Color.color(value)!`? Shouldn't this raise a `SyntaxError: Unexpected token`?

Reply:
`d3Color.color(value)!.darker()`

Reply:
Sorry, this does not make sense to me

Reply:
Just for the record, this is a TypeScript operator: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator